### PR TITLE
Debug tools to help reset settings, delete background options and delete catalog products.

### LINKED ISSUE #2129: Unlinked products lingering in Facebook catalog
## ✍️ Describe the bug

It's possible to have unlinked or “shadow” products in the Facebook catalog which used to be synced but are no longer linked to the WooCommerce site. The IDs of the products on FB don’t match anything in the WooCommerce DB. These products might be a duplicate of an actively-synced product or they may be now-deleted products in WooCommerce. 

In at least one instance (4419208-zen), the shadow products came from the [bug where Draft products were being synced to the FB catalog](https://github.com/woocommerce/facebook-for-woocommerce/issues/1803).

When attempting to delete these products on Facebook, they get this message or a variation of it:
> Permanently remove this item from the catalogue. To delete this item, remove it in the same way that it was added.

Meaning, Facebook is directing the merchant to delete the product from WooCommerce, but there's nothing in WooCommerce to delete.

## 🚶‍♀️ Steps to reproduce

Difficult to reproduce; no known steps at this time.

I've tried reproducing this by disconnecting the Facebook for WooCommerce extension and reconnecting multiple times, deleting products in between, but the products are correctly deleted/updated/synced every time.

### PR DESCRIPTION
**Pull request by rawdreeg**
### Changes proposed in this Pull Request:

Closes #2129.

To help users debug issues with the extension, I have added a few tools to extensions to facilitate some of the common debugging tasks:

- Delete all FB settings data -- This disconnects the FB account and delete all FB settings options.
- Delete Background options job -- Deletes wc_facebook_background_product_sync job options filling up the DB.
- Reset all FB products -- Removes all FB metadata from products so that the product can be synced again.
- Delete all FB products -- Deletes all products in the WooCommerce store from the FB catalog.

- [ x] Do the changed files pass `phpcs` checks? Please remove `phpcs:ignore` comments in changed files and fix any issues, or delete if not practical.

### Screenshots:

Screenshot 2022-11-16 at 16 14 15

### Detailed test instructions:

1. Enable Debug mode: Marketing -> Facebook
2. Then navigate to Tools (WooCommerce -> Status -> Tools) then test the tools.

#### Facebook: Reset connection settings: Should disconnect your FB connection.
1. Run this tool
2. Go to the Facebook admin screen (Marketing -> Facebook), and observe FB has been disconnected
3. Navigate to /wp-admin/options.php. The following options should be empty:
- wc_facebook_access_token
- wc_facebook_merchant_access_token
- wc_facebook_page_access_token
- wc_facebook_commerce_manager_id
- wc_facebook_system_user_id
- wc_facebook_ad_account_id
- wc_facebook_business_manager_id
- wc_facebook_instagram_business_id
- wc_facebook_page_id
- wc_facebook_pixel_id
- wc_facebook_product_catalog_id

#### Facebook: Delete Background Sync Jobs
1. Confirm Facebook is connected and product have been sync to your FB catalog
2. Run this tool
3. Navigate to /wp-admin/options.php. Confirm you have no wc_facebook_background_product_sync

#### Facebook: Reset all products
1. Confirm Facebook is connected and products have been synced to your FB catalog
2. Run this tool
3. Go to the scheduled actions tab and confirm the `facebook_for_woocommerce/jobs/reset_all_catalog_products/chain_start` is pending. Once the job is completed
4. Query the postmeta table using a MySQL client (phpmyadmin). And confirm post meta table does not contain any value for the meta key: fb_product_item_id:
eg: SELECT * FROM `wp_postmeta` WHERE `meta_key` LIKE 'fb_product_item_id'

#### Facebook: Delete all products from your Facebook Catalog
1. Confirm Facebook is connected and products have been synced to your FB catalog
2. Run this tool
3. 3. Go to the scheduled actions tab and confirm the `facebook_for_woocommerce/jobs/delete_products_from_FB_catalog/chain_start` is pending. Once the job is completed
4. Check your FB catalog and confirm products have been deleted.

### Additional details:

1. The tool to delete/ reset products might take some time, depending on the number of products in your catalog.
2. The tool to delete catalog products only deletes products in the WooCommerce store due to limitations in the Meta graph API. Products added from other sources or currently unlinked cannot be deleted by this tool. User might need to contact Meta support to have these deleted.

### Changelog entry

>  Add - Debug tools to help reset settings, delete background options and delete catalog products.

### CODE REVIEW

File: includes/Utilities/DebugTools.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+		$tools['wc_facebook_delete_background_jobs'] = [
+			'name'     => __( 'Facebook: Delete Background Sync Jobs', 'facebook-for-woocommerce' ),
+			'button'   => __( 'Clear Background Sync Jobs', 'facebook-for-woocommerce' ),
+			'desc'     => __( 'This tool will clear your clear background sync jobs from the options table.', 'facebook-for-woocommerce' ),
+			'callback' => [ $this, 'clean_up_old_background_sync_options' ],
+		];
+
+		$tools['reset_all_product_fb_settings'] = [
+			'name'     => __( 'Facebook: Reset all products', 'facebook-for-woocommerce' ),
```
[the remark this answers is not in the document]
Comment (rawdreeg) [reply includes/Utilities/DebugTools.php:52]: Yes. Reset only removes FB metadata associated with the woo product. Does not remove the product from FB at all. You're welcome to suggest a better name.

File: class-wc-facebookcommerce.php [hunk 2]
```diff
@@ -182,6 +186,10 @@ public function init() {
 			$this->job_manager = new WooCommerce\Facebook\Jobs\JobManager();
 			add_action( 'init', [ $this->job_manager, 'init' ] );
 
+			// Load debug tools if debug mode is enabled.
+			require_once __DIR__ . '/includes/Utilities/DebugTools.php';
```
[the remark this answers is not in the document]
Comment (rawdreeg) [reply class-wc-facebookcommerce.php]: Yes, you're correct this autoload. I removed the required_one statement in c0d0953efd01e53e5141826f1c017615749a814d

File: includes/Utilities/DebugTools.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	/**
+	 * Runs the clear settings tool.
+	 *
+	 * @since x.x.x
+	 *
+	 * @return string
+	 */
+	public function clear_facebook_settings() {
+		// Disconnect FB.
+		facebook_for_woocommerce()->get_connection_handler()->disconnect();
```
[the remark this answers is not in the document]
Comment (rawdreeg) [reply includes/Utilities/DebugTools.php:93]: Updated the disconnect method to also clear the page_access_token in 24e9f538c1a1ecaae696acbca333ee61c3112615. Thanks for catching this.

File: includes/Jobs/DeleteProductsFromFBCatalog.php [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	 *
+	 * @throws Exception On error. The failure will be logged by Action Scheduler and the job chain will stop.
+	 */
+	protected function process_items( array $items, array $args ) {
+		$integration = facebook_for_woocommerce()->get_integration();
+		foreach ( $items as $product_id ) {
+			$product = wc_get_product( $product_id );
+			// check if variable product
+			if ( $product->is_type( 'variable' ) ) {
+				$integration->delete_product_group( $product_id );
```
[the remark this answers is not in the document]
Comment (rawdreeg) [reply includes/Jobs/DeleteProductsFromFBCatalog.php:79]: This has been addressed in df1d70094d8a455b75f0434375b41acf69681708. Thanks for catching this.

File: includes/Utilities/DebugTools.php [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	/**
+	 * Runs the reset all catalog products settings tool.
+	 *
+	 * @since x.x.x
+	 *
+	 * @return string
+	 */
+	public function reset_all_product_fb_settings() {
```
[the remark this answers is not in the document]
Comment (rawdreeg) [reply includes/Utilities/DebugTools.php:106]: This should start with a cron execution. Or can be triggered manually

File: includes/Utilities/DebugTools.php [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	/**
+	 * Delete products from Facebook catalog.
+	 *
+	 * @since x.x.x
+	 *
+	 * @return string
+	 */
+	public function delete_all_products() {
```
[the remark this answers is not in the document]
Comment (rawdreeg) [reply includes/Utilities/DebugTools.php:119]: Addressed in df1d70094d8a455b75f0434375b41acf69681708. Thanks

**Comment by rawdreeg** [comment]
@ibndawood, I updated the testing instruction for the Reset All products tool.

**Comment by rawdreeg** [comment]
@ibndawood This is ready for another round of review.

Error when updating after delete has been solved by: 8c641d05061e9b8e8562da0faa7229266f94fa4f

And I updated the instruction for **Facebook: Reset all products** testing.

Thanks

**Comment by paigeql** [comment]
Is there seller-friendly documentation on this? We have a lot of clients to whom we have to explain how deletion logic works with the Facebook <> WC integration so this fix is actually quite siginificant to support.

**Comment by rawdreeg** [comment]
Hi @paigeql, we updated the Facebook for WooCommerce with documentation: https://woocommerce.com/document/facebook-for-woocommerce/#section-26.

Let us know if this helps. 

Thanks, R

**Pull request merged**